### PR TITLE
Implement CQuorum and CQuorumManager

### PR DESCRIPTION
**Pull request by codablock**

### CODE REVIEW

File: src/llmq/quorums.cpp [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+    auto members = CLLMQUtils::GetAllQuorumMembers((Consensus::LLMQType)qc.llmqType, qc.quorumHash);
+
+    quorum->Init(qc.quorumHash, quorumIndex->nHeight, members, qc.validMembers, qc.quorumPublicKey);
+
+    if (!quorum->ReadContributions(evoDb)) {
+        if (BuildQuorumContributions(qc, quorum)) {
+            quorum->WriteContributions(evoDb);
+        } else {
+            LogPrintf("CQuorumManager::%s -- quorum.ReadContributions and BuildQuorumContributions for block %s failed", __func__, qc.quorumHash.ToString());
+        }
```
Comment (UdjinM6) [inline src/llmq/quorums.cpp]:
```suggestion
            return false;
        }
```

File: src/llmq/quorums.cpp [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+    }
+
+    BLSVerificationVectorPtr quorumVvec;
+    CBLSSecretKey skShare;
+
+    cxxtimer::Timer t2(true);
+    quorumVvec = blsWorker.BuildQuorumVerificationVector(vvecs);
+    if (quorumVvec == nullptr) {
+        LogPrintf("CQuorumManager::%s -- failed to build quorumVvec\n", __func__);
+    }
```
Comment (UdjinM6) [inline src/llmq/quorums.cpp]:
```suggestion
        return false;
    }
```

File: src/llmq/quorums.cpp [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+    skShare = blsWorker.AggregateSecretKeys(skContributions);
+    if (!skShare.IsValid()) {
+        LogPrintf("CQuorumManager::%s -- failed to build skShare\n", __func__);
+    }
```
Comment (UdjinM6) [inline src/llmq/quorums.cpp]:
```suggestion
        return false;
    }
```

File: src/rpc/rpcquorums.cpp [hunk 4]
```diff
@@ -2,12 +2,116 @@
...
+#include "llmq/quorums.h"
 #include "llmq/quorums_debug.h"
 #include "llmq/quorums_dkgsession.h"
 
+void quorum_list_help()
+{
+    throw std::runtime_error(
+            "quorum list (count)\n"
+            "\nArguments:\n"
+            "1. count           (number, optional) Number of quorums to list.\n"
```
Comment (UdjinM6) [inline src/rpc/rpcquorums.cpp]:
```suggestion
            "1. count           (number, optional, default=10) Number of quorums to list.\n"
```

File: src/rpc/rpcquorums.cpp [hunk 5]
```diff
@@ -2,12 +2,116 @@
...
+    }
+
+
+    return ret;
+}
+
+void quorum_info_help()
+{
+    throw std::runtime_error(
+            "quorum info \"llmqType\" \"blockquorumHash\" (includeSkShare)\n"
```
Comment (UdjinM6) [inline src/rpc/rpcquorums.cpp]:
```suggestion
            "quorum info llmqType \"quorumHash\" ( includeSkShare )\n"
```

File: src/rpc/rpcquorums.cpp [hunk 6]
```diff
@@ -2,12 +2,116 @@
...
+            "\nArguments:\n"
+            "1. \"llmqType\"            (int, required) LLMQ type.\n"
```
Comment (UdjinM6) [inline src/rpc/rpcquorums.cpp]:
```suggestion
            "1. llmqType              (int, required) LLMQ type.\n"
```

File: src/rpc/rpcquorums.cpp [hunk 7]
```diff
@@ -2,12 +2,116 @@
...
+            "2. \"quorumHash\"          (string, required) Block hash of quorum.\n"
+            "3. \"includeSkShare\"      (boolean, optional) Include secret key share in output.\n"
```
Comment (UdjinM6) [inline src/rpc/rpcquorums.cpp]:
```suggestion
            "3. includeSkShare        (boolean, optional, default=false) Include secret key share in output.\n"
```

File: src/rpc/rpcquorums.cpp [hunk 8]
```diff
@@ -2,12 +2,116 @@
...
+
+    uint256 blockHash = ParseHashV(request.params[2], "quorumHash");
+    bool includeSkShare = false;
+    if (request.params.size() > 3) {
+        includeSkShare = ParseBoolV(request.params[3], "includeSkShare");
+    }
+
+    auto quorum = llmq::quorumManager->GetQuorum(llmqType, blockHash);
+    if (!quorum) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "quorum not found");
```
Comment (UdjinM6) [inline src/rpc/rpcquorums.cpp]:
```suggestion
        throw JSONRPCError(RPC_INTERNAL_ERROR, "quorum not found");
```

File: src/rpc/rpcquorums.cpp [hunk 9]
```diff
@@ -61,7 +165,11 @@ UniValue quorum(const JSONRPCRequest& request)
 
     std::string command = request.params[0].get_str();
 
-    if (command == "dkgstatus") {
+    if (command == "list") {
```
Comment (UdjinM6) [inline src/rpc/rpcquorums.cpp]: Should probably add `quorum_help()` with the list of available commands, smth similar to `protx_help()`, and also unify the logic flow here with the one in `protx()` (and other "complex" rpc commands).

File: src/llmq/quorums.cpp (hunk 2 above)
Comment (codablock) [reply src/llmq/quorums.cpp]: Same. Added a comment.

File: src/rpc/rpcquorums.cpp (hunk 8 above)
Comment (codablock) [reply src/rpc/rpcquorums.cpp]: why `RPC_INTERNAL_ERROR` and not `RPC_INVALID_PARAMETER`?

File: src/rpc/rpcquorums.cpp (hunk 9 above)
Comment (codablock) [reply src/rpc/rpcquorums.cpp]: done

File: src/llmq/quorums.cpp [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+}
+
+bool CQuorum::WriteContributions(CEvoDB& evoDb)
+{
+    uint256 dbKey = MakeQuorumKey(*this);
+
+    if (quorumVvec != nullptr) {
+        evoDb.GetRawDB().Write(std::make_pair(DB_QUORUM_QUORUM_VVEC, dbKey), *quorumVvec);
+    }
+    if (skShare.IsValid()) {
```
Comment (UdjinM6) [inline src/llmq/quorums.cpp]: Is it possible to have valid `skShare` without having `quorumVvec`? Also, shouldn't we fail if `quorumVvec == nullptr`? 
I.e. maybe
```
bool CQuorum::WriteContributions(CEvoDB& evoDb)
{
    uint256 dbKey = MakeQuorumKey(*this);

    if (quorumVvec == nullptr) return false;

    evoDb.GetRawDB().Write(std::make_pair(DB_QUORUM_QUORUM_VVEC, dbKey), *quorumVvec);

    if (skShare.IsValid()) {
        evoDb.GetRawDB().Write(std::make_pair(DB_QUORUM_SK_SHARE, dbKey), skShare);
    }

    return true;
}
```
?

Comment (codablock) [reply src/llmq/quorums.cpp]: No, without a `quorumVvec` there can't be a `skShare`. Maybe I should completely remove the return value of this method as it doesn't really fail when it doesn't write anything. Also, the return value of this method is actually never used.

Comment (UdjinM6) [reply src/llmq/quorums.cpp]: Yeah, returning `bool` for `Write...` methods is probably not that useful.

Comment (codablock) [reply src/llmq/quorums.cpp]: Changed it to use `void`. Need a re-ACK now

**Comment by codablock** [comment]
Hmm, test failure (no log output from unit tests) happened the second time on this branch already. Other times it worked. Not sure if we can ignore this.

**Comment by UdjinM6** [comment]
Yeah, that's concerning a bit but I don't see how any of recent changes could affect wallet tests, it's probably some older bug/issue.

**Pull request merged**